configs/provider.go

```diff
@@ -107,28 +108,82 @@ func (p *Provider) moduleUniqueKey() string {
 	return p.Name
 }
 
-// ProviderRequirement represents a declaration of a dependency on a particular
-// provider version without actually configuring that provider. This is used in
-// child modules that expect a provider to be passed in from their parent.
-type ProviderRequirement struct {
-	Name        string
-	Requirement VersionConstraint
+// ParseProviderConfigCompact parses the given absolute traversal as a relative
+// provider address in compact form. The following are examples of traversals
+// that can be successfully parsed as compact relative provider configuration
+// addresses:
+//
+//     aws
+//     aws.foo
+//
+// This function will panic if given a relative traversal.
+//
+// If the returned diagnostics contains errors then the result value is invalid
+// and must not be used.
+func ParseProviderConfigCompact(traversal hcl.Traversal) (addrs.ProviderConfig, tfdiags.Diagnostics) {
```

Comment:
Why is this moving to the configs package? It seems to only be directly handling hcl and addrs types.

Reply:
Ah, sorry, that should have been in the commit message 🙄 

This change is based on a conversation with @apparentlymart, partially offline and partially in this PR: https://github.com/hashicorp/terraform/pull/23575#discussion_r354566008

The idea is that it no longer makes any sense to parse a provider configuration reference outside the context of the configuration language: you can't parse `aws.foo` into `"registry.terraform.io/hashicorp/aws alias foo"` because it's expressed in a context where we need to complete the local name lookup first. 

This change merely moves the function, while the next step is to add a new type that encapsulates "reference to a provider config using its local name within a single module", and have this function return that type. 

I'm not opposed to reverting this particular change for the time being; I included it here as a baby-step but on reflection it may be too small of a step to make sense in isolation.

Reply:
Thanks for the extra context. I'm OK with this as long as this is an intentional intermediate step to to handle the new provider information.